README
NAME
    Coro - the only real threads in perl

SYNOPSIS
      use Coro;
  
      async {
         # some asynchronous thread of execution
         print "2\n";
         cede; # yield back to main
         print "4\n";
      };
      print "1\n";
      cede; # yield to coro
      print "3\n";
      cede; # and again
  
      # use locking
      my $lock = new Coro::Semaphore;
      my $locked;
  
      $lock->down;
      $locked = 1;
      $lock->up;

DESCRIPTION
    For a tutorial-style introduction, please read the Coro::Intro manpage.
    This manpage mainly contains reference information.

    This module collection manages continuations in general, most often in
    the form of cooperative threads (also called coros, or simply "coro" in
    the documentation). They are similar to kernel threads but don't (in
    general) run in parallel at the same time even on SMP machines. The
    specific flavor of thread offered by this module also guarantees you
    that it will not switch between threads unless necessary, at
    easily-identified points in your program, so locking and parallel access
    are rarely an issue, making thread programming much safer and easier
    than using other thread models.

    Unlike the so-called "Perl threads" (which are not actually real threads
    but only the windows process emulation (see section of same name for
    more details) ported to UNIX, and as such act as processes), Coro
    provides a full shared address space, which makes communication between
    threads very easy. And coro threads are fast, too: disabling the Windows
    process emulation code in your perl and using Coro can easily result in
    a two to four times speed increase for your programs. A parallel matrix
    multiplication benchmark (very communication-intensive) runs over 300
    times faster on a single core than perls pseudo-threads on a quad core
    using all four cores.

    Coro achieves that by supporting multiple running interpreters that
    share data, which is especially useful to code pseudo-parallel processes
    and for event-based programming, such as multiple HTTP-GET requests
    running concurrently. See Coro::AnyEvent to learn more on how to
    integrate Coro into an event-based environment.

    In this module, a thread is defined as "callchain + lexical variables +
    some package variables + C stack), that is, a thread has its own
    callchain, its own set of lexicals and its own set of perls most
    important global variables (see Coro::State for more configuration and
    background info).

    See also the "SEE ALSO" section at the end of this document - the Coro
    module family is quite large.

CORO THREAD LIFE CYCLE
    During the long and exciting (or not) life of a coro thread, it goes
    through a number of states:

    1. Creation
        The first thing in the life of a coro thread is it's creation -
        obviously. The typical way to create a thread is to call the "async
        BLOCK" function:

           async {
              # thread code goes here
           };

        You can also pass arguments, which are put in @_:

           async {
              print $_[1]; # prints 2
           } 1, 2, 3;

        This creates a new coro thread and puts it into the ready queue,
        meaning it will run as soon as the CPU is free for it.

        "async" will return a Coro object - you can store this for future
        reference or ignore it - a thread that is running, ready to run or
        waiting for some event is alive on it's own.

        Another way to create a thread is to call the "new" constructor with
        a code-reference:

           new Coro sub {
              # thread code goes here
           }, @optional_arguments;

        This is quite similar to calling "async", but the important
        difference is that the new thread is not put into the ready queue,
        so the thread will not run until somebody puts it there. "async" is,
        therefore, identical to this sequence:

           my $coro = new Coro sub {
              # thread code goes here
           };
           $coro->ready;
           return $coro;

    2. Startup
        When a new coro thread is created, only a copy of the code reference
        and the arguments are stored, no extra memory for stacks and so on
        is allocated, keeping the coro thread in a low-memory state.

        Only when it actually starts executing will all the resources be
        finally allocated.

        The optional arguments specified at coro creation are available in
        @_, similar to function calls.

    3. Running / Blocking
        A lot can happen after the coro thread has started running. Quite
        usually, it will not run to the end in one go (because you could use
        a function instead), but it will give up the CPU regularly because
        it waits for external events.

        As long as a coro thread runs, its Coro object is available in the
        global variable $Coro::current.

        The low-level way to give up the CPU is to call the scheduler, which
        selects a new coro thread to run:

           Coro::schedule;

        Since running threads are not in the ready queue, calling the
        scheduler without doing anything else will block the coro thread
        forever - you need to arrange either for the coro to put woken up
        (readied) by some other event or some other thread, or you can put
        it into the ready queue before scheduling:

           # this is exactly what Coro::cede does
           $Coro::current->ready;
           Coro::schedule;

        All the higher-level synchronisation methods (Coro::Semaphore,
        Coro::rouse_*...) are actually implemented via "->ready" and
        "Coro::schedule".

        While the coro thread is running it also might get assigned a
        C-level thread, or the C-level thread might be unassigned from it,
        as the Coro runtime wishes. A C-level thread needs to be assigned
        when your perl thread calls into some C-level function and that
        function in turn calls perl and perl then wants to switch
        coroutines. This happens most often when you run an event loop and
        block in the callback, or when perl itself calls some function such
        as "AUTOLOAD" or methods via the "tie" mechanism.

    4. Termination
        Many threads actually terminate after some time. There are a number
        of ways to terminate a coro thread, the simplest is returning from
        the top-level code reference:

           async {
              # after returning from here, the coro thread is terminated
           };

           async {
              return if 0.5 <  rand; # terminate a little earlier, maybe
              print "got a chance to print this\n";
              # or here
           };

        Any values returned from the coroutine can be recovered using
        "->join":

           my $coro = async {
              "hello, world\n" # return a string
           };

           my $hello_world = $coro->join;

           print $hello_world;

        Another way to terminate is to call "Coro::terminate", which at any
        subroutine call nesting level:

           async {
              Coro::terminate "return value 1", "return value 2";
           };

        Yet another way is to "->cancel" (or "->safe_cancel") the coro
        thread from another thread:

           my $coro = async {
              exit 1;
           };

           $coro->cancel; # also accepts values for ->join to retrieve

        Cancellation *can* be dangerous - it's a bit like calling "exit"
        without actually exiting, and might leave C libraries and XS modules
        in a weird state. Unlike other thread implementations, however, Coro
        is exceptionally safe with regards to cancellation, as perl will
        always be in a consistent state, and for those cases where you want
        to do truly marvellous things with your coro while it is being
        cancelled - that is, make sure all cleanup code is executed from the
        thread being cancelled - there is even a "->safe_cancel" method.

        So, cancelling a thread that runs in an XS event loop might not be
        the best idea, but any other combination that deals with perl only
        (cancelling when a thread is in a "tie" method or an "AUTOLOAD" for
        example) is safe.

        Last not least, a coro thread object that isn't referenced is
        "->cancel"'ed automatically - just like other objects in Perl. This
        is not such a common case, however - a running thread is referencedy
        by $Coro::current, a thread ready to run is referenced by the ready
        queue, a thread waiting on a lock or semaphore is referenced by
        being in some wait list and so on. But a thread that isn't in any of
        those queues gets cancelled:

           async {
              schedule; # cede to other coros, don't go into the ready queue
           };

           cede;
           # now the async above is destroyed, as it is not referenced by anything.

        A slightly embellished example might make it clearer:

           async {
              my $guard = Guard::guard { print "destroyed\n" };
              schedule while 1;
           };

           cede;

        Superficially one might not expect any output - since the "async"
        implements an endless loop, the $guard will not be cleaned up.
        However, since the thread object returned by "async" is not stored
        anywhere, the thread is initially referenced because it is in the
        ready queue, when it runs it is referenced by $Coro::current, but
        when it calls "schedule", it gets "cancel"ed causing the guard
        object to be destroyed (see the next section), and printing it's
        message.

        If this seems a bit drastic, remember that this only happens when
        nothing references the thread anymore, which means there is no way
        to further execute it, ever. The only options at this point are
        leaking the thread, or cleaning it up, which brings us to...

    5. Cleanup
        Threads will allocate various resources. Most but not all will be
        returned when a thread terminates, during clean-up.

        Cleanup is quite similar to throwing an uncaught exception: perl
        will work it's way up through all subroutine calls and blocks. On
        it's way, it will release all "my" variables, undo all "local"'s and
        free any other resources truly local to the thread.

        So, a common way to free resources is to keep them referenced only
        by my variables:

           async {
              my $big_cache = new Cache ...;
           };

        If there are no other references, then the $big_cache object will be
        freed when the thread terminates, regardless of how it does so.

        What it does "NOT" do is unlock any Coro::Semaphores or similar
        resources, but that's where the "guard" methods come in handy:

           my $sem = new Coro::Semaphore;

           async {
              my $lock_guard = $sem->guard;
              # if we return, or die or get cancelled, here,
              # then the semaphore will be "up"ed.
           };

        The "Guard::guard" function comes in handy for any custom cleanup
        you might want to do (but you cannot switch to other coroutines from
        those code blocks):

           async {
              my $window = new Gtk2::Window "toplevel";
              # The window will not be cleaned up automatically, even when $window
              # gets freed, so use a guard to ensure it's destruction
              # in case of an error:
              my $window_guard = Guard::guard { $window->destroy };

              # we are safe here
           };

        Last not least, "local" can often be handy, too, e.g. when
        temporarily replacing the coro thread description:

           sub myfunction {
              local $Coro::current->{desc} = "inside myfunction(@_)";

              # if we return or die here, the description will be restored
           }

    6. Viva La Zombie Muerte
        Even after a thread has terminated and cleaned up its resources, the
        Coro object still is there and stores the return values of the
        thread.

        When there are no other references, it will simply be cleaned up and
        freed.

        If there areany references, the Coro object will stay around, and
        you can call "->join" as many times as you wish to retrieve the
        result values:

           async {
              print "hi\n";
              1
           };

           # run the async above, and free everything before returning
           # from Coro::cede:
           Coro::cede;

           {
              my $coro = async {
                 print "hi\n";
                 1
              };

              # run the async above, and clean up, but do not free the coro
              # object:
              Coro::cede;

              # optionally retrieve the result values
              my @results = $coro->join;

              # now $coro goes out of scope, and presumably gets freed
           };

GLOBAL VARIABLES
    $Coro::main
        This variable stores the Coro object that represents the main
        program. While you can "ready" it and do most other things you can
        do to coro, it is mainly useful to compare again $Coro::current, to

Changes
Revision history for Perl extension Coro.

TODO: should explore PerlIO::coroaio (perl leaks like hell).
TODO: channel->maxsize(newsize)?
TODO: __GCC_HAVE_DWARF2_CFI_ASM
TODO: swap_sv, maybe add scope_swap_sv?
TODO: croak when async_pool tries to run canceled thread?
TODO: how to debug safe_cancle failure (vebrose mode? show how to make backtrace?)

	- minor code cleanups.

6.514 Thu Aug 31 18:25:31 CEST 2017
	- safe_cancel no longer croaks when called on already-destroyed
          thread, but instead "just works".

6.513 Sat Jul 15 01:18:51 CEST 2017
	- experimental 5.22 comaptibility changes (PadlistNAMES).
	- removed attributes example and made sure the other examples
          still work.
        - fix some statevar types for newer perls, should be of no consequence.

6.512 Fri Jul 14 05:03:17 CEST 2017
        - seems to pass test suite down till 5.8.9, so make this the
          minimum requirement in Makefile.PL.
	- experimental 5.24 compatibility changes (SUB_ARGARRAY).
        - experimental 5.26 compatibility changes (PL_comppad -> PADOFFSET).
        - disable FORTIFY_SOURCE to avoid broken (but well-intentioned)
          fortified longjmp variants in some libcs.
	- re-try JIT allocation without PROT_EXEC, to hopefully improve
          portability (still doesn't seem to work around broken selinux
          mmap everywhere).
        - speling fixes (Xavier Guimard).

6.511 Sun Jun 26 23:44:50 CEST 2016
	- make it compile with threaded perls again.
	- simplify and speed up __DIE__ and __WARN__ handling.

6.51 Sat Jun 25 20:57:20 CEST 2016
	- 6.5 release didn't compile on most older perls due to
          a logic error enabling experimental 5.24 code for all versions.
        - since various broken or deficient patches float around
          to work around the vtbl API breakage in perl, let's provide
          our own workaround, which seems to be much less invasive
          and more compatible than the approaches seen so far. YMMV.
	- work around assertion failure in perl_destruct on 5.24
          (analyzed by Dave Mitchell).

6.5  Wed Jun 22 22:23:50 CEST 2016
	- swap_sv swapping order was not symmetrical, causing
          wrong swaps when swap_sv was used multiple times on the
          same sv in the same thread.
        - swap_sv calls can now be undone by calling it again
          with the same variables.
        - swap_sv calls will now be undone in async_pool threads.
	- split Coro::Semaphore::up/adjust into separate xs functions
          for better error reporting, at slight codesize increase.
        - (libcoro) arm assembly support, please test and report.
        - adjust to PL_savestack changes in perl 5.24 (adapted from
          the debian patch, which unfortunately gets it wrong).

6.49 Sat Oct 17 01:40:12 CEST 2015
	- throwing an exception to a thread waiting in
          Coro::Handle using Coro::EV did not stop the watchers,
          causing the next call to fail (testcase by Martin Pritchard).
	- bump minimum perl version to 5.10.

6.48 Sun Oct  4 19:03:51 CEST 2015
	- fix memory leak when Coro::Handle uses Coro::EV internally
          (testcase by Sten Sten).
        - update code to libev 4 API (internally, Coro still used the
          version 3 API calls).
        - remove unused hv_sig.

6.47 Sat Jul 11 03:58:20 CEST 2015
	- swap_svs - NVs can be in the SV head beginning with 5.20.

6.46 Tue Jun 30 14:36:30 CEST 2015
	- restore portability to perls with windows fork emulation
          (patch by Petr Písař).

6.45 Tue Jun 30 01:40:08 CEST 2015
	- restore portability to perl 5.16 and below.

6.44 Tue Jun 30 00:41:54 CEST 2015
	- avoid segfaulting (or worse) when tracing a canceled thread.
	- the Event module silently broke it's public hook API, causing
          failures on perls compiled with -Duselongdouble. Requiring
          the latest version of Event because I don't know which release
          changed this (it's not mentioned in the ChangeLog).
        - allow xs level enterleave hooks via CoroAPI.

6.43 Thu Jun  4 15:38:14 CEST 2015
	- use stability canary.
	- port to stableperl-5.22.0-1.001.
        - update libecb.

6.42 Wed Feb 11 20:29:52 CET 2015
	- Coro::SemaphoreSet->try did not actually work (analyzed by
          SATO Kentaro).
        - upgrade libecb to be C11 compliant.

6.41 Sat Sep  6 22:08:46 CEST 2014
	- restore portability to perl 5.8.x.
        - give new Coro's a valid GvHV(PL_hintgv) - this is slow and takes
          up some memory, but fixes "use feature" and similar modules
          when used inside a Coro.
        - allow zero as argument to Coro::Channel to mean the same thing
          as no argument. this works with older versions as well,
          but wasn't legal until now.
        - slightly better c header file detection.

6.39 Mon Jun  2 00:00:08 CEST 2014
	- work around more incompatible changes in 5.20.

6.38 Sun Jun  1 21:54:23 CEST 2014
	- check that perl slots actually have enough space to hold
          interpreter variables.
	- untested port to perl 5.19 (64 bit tmps indices) (reported
          by Andreas König).
	- croak when cancel is called without a thread context.

6.37 Tue Mar  4 13:27:33 CET 2014
	- *sigh*, removed leftover debugging code from debugging a
          perl bug, of all things.

6.36 Tue Mar  4 07:11:59 CET 2014
	- semaphores would not clear the destroy hook when interrupted
          by ->throw, causing segfaults or worse.
        - ->throw on a thread waiting for a semaphore did not acquire
          the semaphore, but also didn't wake up other waiters,
          possibly causing a deadlock.
	- "FATAL: $Coro::IDLE blocked itself" will now use Carp::confess
          to report a full stacktrace, which should help find out
          where the actual call is.
        - "atomically" destroy data in slf_destroy, because it is
          the right thing to do, just in case.
        - disable ecb.h libm dependency, to compile on ancient systems
          or under adverse conditions.


Coro/Debug.pm
=head1 NAME

Coro::Debug - various functions that help debugging Coro programs

=head1 SYNOPSIS

 use Coro::Debug;

 our $server = new_unix_server Coro::Debug "/tmp/socketpath";

 $ socat readline unix:/tmp/socketpath

=head1 DESCRIPTION

This module is an L<AnyEvent> user, you need to make sure that you use and
run a supported event loop.

This module provides some debugging facilities. Most will, if not handled
carefully, severely compromise the security of your program, so use it
only for debugging (or take other precautions).

It mainly implements a very primitive debugger that is very easy to
integrate in your program:

   our $server = new_unix_server Coro::Debug "/tmp/somepath";
   # see new_unix_server, below, for more info

It lets you list running coroutines:

            state (rUnning, Ready, New or neither)
            |cctx allocated
            ||  resident set size (octets)
            ||  |   scheduled this many times
   > ps     ||  |   |
        PID SC  RSS USES Description              Where
   14572344 UC  62k 128k [main::]                 [dm-support.ext:47]
   14620056 -- 2260   13 [coro manager]           [Coro.pm:358]
   14620128 -- 2260  166 [unblock_sub scheduler]  [Coro.pm:358]
   17764008 N-  152    0 [EV idle process]        -
   13990784 -- 2596  10k timeslot manager         [cf.pm:454]
   81424176 --  18k 4758 [async pool idle]        [Coro.pm:257]
   23513336 -- 2624    1 follow handler           [follow.ext:52]
   40548312 --  15k 5597 player scheduler         [player-scheduler.ext:13]
   29138032 -- 2548  431 music scheduler          [player-env.ext:77]
   43449808 -- 2260 3493 worldmap updater         [item-worldmap.ext:115]
   33352488 --  19k 2845 [async pool idle]        [Coro.pm:257]
   81530072 --  13k  43k map scheduler            [map-scheduler.ext:65]
   30751144 --  15k 2204 [async pool idle]        [Coro.pm:257]

Lets you do backtraces on about any coroutine:

   > bt 18334288
   coroutine is at /opt/cf/ext/player-env.ext line 77
           eval {...} called at /opt/cf/ext/player-env.ext line 77
           ext::player_env::__ANON__ called at -e line 0
           Coro::_run_coro called at -e line 0

Or lets you eval perl code:

   > 5+7
   12

Or lets you eval perl code within other coroutines:

   > eval 18334288 caller(1); $DB::args[0]->method
   1

It can also trace subroutine entry/exits for most coroutines (those not
having recursed into a C function), resulting in output similar to:

   > loglevel 5
   > trace 94652688
   2007-09-27Z20:30:25.1368 (5) [94652688] enter Socket::sockaddr_in with (8481,\x{7f}\x{00}\x{00}\x{01})
   2007-09-27Z20:30:25.1369 (5) [94652688] leave Socket::sockaddr_in returning (\x{02}\x{00}...)
   2007-09-27Z20:30:25.1370 (5) [94652688] enter Net::FCP::Util::touc with (client_get)
   2007-09-27Z20:30:25.1371 (5) [94652688] leave Net::FCP::Util::touc returning (ClientGet)
   2007-09-27Z20:30:25.1372 (5) [94652688] enter AnyEvent::Impl::Event::io with (AnyEvent,fh,GLOB(0x9256250),poll,w,cb,CODE(0x8c963a0))
   2007-09-27Z20:30:25.1373 (5) [94652688] enter Event::Watcher::__ANON__ with (Event,poll,w,fd,GLOB(0x9256250),cb,CODE(0x8c963a0))
   2007-09-27Z20:30:25.1374 (5) [94652688] enter Event::io::new with (Event::io,poll,w,fd,GLOB(0x9256250),cb,CODE(0x8c963a0))
   2007-09-27Z20:30:25.1375 (5) [94652688] enter Event::Watcher::init with (Event::io=HASH(0x8bfb120),HASH(0x9b7940))

If your program uses the Coro::Debug::log facility:

   Coro::Debug::log 0, "important message";
   Coro::Debug::log 9, "unimportant message";

Then you can even receive log messages in any debugging session:

   > loglevel 5
   2007-09-26Z02:22:46 (9) unimportant message

Other commands are available in the shell, use the C<help> command for a list.

=head1 FUNCTIONS

None of the functions are being exported.

=over 4

=cut

package Coro::Debug;

use common::sense;

use overload ();

use Carp ();
use Scalar::Util ();

use Guard;

use AnyEvent ();
use AnyEvent::Util ();
use AnyEvent::Socket ();

use Coro ();
use Coro::Handle ();
use Coro::State ();
use Coro::AnyEvent ();
use Coro::Timer ();

our $VERSION = 6.512;

our %log;
our $SESLOGLEVEL = exists $ENV{PERL_CORO_DEFAULT_LOGLEVEL} ? $ENV{PERL_CORO_DEFAULT_LOGLEVEL} : -1;
our $ERRLOGLEVEL = exists $ENV{PERL_CORO_STDERR_LOGLEVEL}  ? $ENV{PERL_CORO_STDERR_LOGLEVEL}  : -1;

sub find_coro {
   my ($pid) = @_;

   if (my ($coro) = grep $_ == $pid, Coro::State::list) {
      $coro
   } else {
      print "$pid: no such coroutine\n";
      undef
   }
}

sub format_msg($$) {
   my ($time, $micro) = Coro::Util::gettimeofday;
   my ($sec, $min, $hour, $day, $mon, $year) = gmtime $time;
   my $date = sprintf "%04d-%02d-%02dZ%02d:%02d:%02d.%04d",
                      $year + 1900, $mon + 1, $day, $hour, $min, $sec, $micro / 100;
   sprintf "%s (%d) %s", $date, $_[0], $_[1]
}

sub format_num4($) {
   my ($v) = @_;

   return sprintf "%4d"   , $v                     if $v <  1e4;
   # 1e5 redundant
   return sprintf "%3.0fk", $v /             1_000 if $v <  1e6;
   return sprintf "%1.1fM", $v /         1_000_000 if $v <  1e7 * .995;
   return sprintf "%3.0fM", $v /         1_000_000 if $v <  1e9;
   return sprintf "%1.1fG", $v /     1_000_000_000 if $v < 1e10 * .995;
   return sprintf "%3.0fG", $v /     1_000_000_000 if $v < 1e12;
   return sprintf "%1.1fT", $v / 1_000_000_000_000 if $v < 1e13 * .995;
   return sprintf "%3.0fT", $v / 1_000_000_000_000 if $v < 1e15;

   "++++"
}

=item log $level, $msg

Log a debug message of the given severity level (0 is highest, higher is
less important) to all interested parties.

=item stderr_loglevel $level

Set the loglevel for logging to stderr (defaults to the value of the
environment variable PERL_CORO_STDERR_LOGLEVEL, or -1 if missing).

=item session_loglevel $level

Set the default loglevel for new coro debug sessions (defaults to the
value of the environment variable PERL_CORO_DEFAULT_LOGLEVEL, or -1 if
missing).

=cut

sub log($$) {
   my ($level, $msg) = @_;
   $msg =~ s/\s*$/\n/;
   $_->($level, $msg) for values %log;
   printf STDERR format_msg $level, $msg if $level <= $ERRLOGLEVEL;
}

sub session_loglevel($) {
   $SESLOGLEVEL = shift;
}

sub stderr_loglevel($) {
   $ERRLOGLEVEL = shift;
}

=item trace $coro, $loglevel

Enables tracing the given coroutine at the given loglevel. If loglevel is
omitted, use 5. If coro is omitted, trace the current coroutine. Tracing
incurs a very high runtime overhead.

It is not uncommon to enable tracing on oneself by simply calling
C<Coro::Debug::trace>.

A message will be logged at the given loglevel if it is not possible to
enable tracing.

=item untrace $coro

Disables tracing on the given coroutine.

=cut

sub trace {
   my ($coro, $loglevel) = @_;

   $coro ||= $Coro::current;
   $loglevel = 5 unless defined $loglevel;

   (Coro::async {
      if (eval { Coro::State::trace $coro, Coro::State::CC_TRACE | Coro::State::CC_TRACE_SUB; 1 }) {
         Coro::Debug::log $loglevel, sprintf "[%d] tracing enabled", $coro + 0;
         $coro->{_trace_line_cb} = sub {
            Coro::Debug::log $loglevel, sprintf "[%d] at %s:%d\n", $Coro::current+0, @_;
         };
         $coro->{_trace_sub_cb} = sub {
            Coro::Debug::log $loglevel, sprintf "[%d] %s %s %s\n",
               $Coro::current+0,
               $_[0] ? "enter" : "leave",
               $_[1],
               $_[2] ? ($_[0] ? "with (" : "returning (") . (
                  join ",",
                     map {
                        my $x = ref $_ ? overload::StrVal $_ : $_;
                        (substr $x, 40) = "..." if 40 + 3 < length $x;
                        $x =~ s/([^\x20-\x5b\x5d-\x7e])/sprintf "\\x{%02x}", ord $1/ge;
                        $x
                     } @{$_[2]}
               ) . ")" : "";
         };

         undef $coro; # the subs keep a reference which we do not want them to do
      } else {
         Coro::Debug::log $loglevel, sprintf "[%d] unable to enable tracing: %s", $Coro::current + 0, $@;
      }
   })->prio (Coro::PRIO_MAX);

   Coro::cede;
}

sub untrace {
   my ($coro) = @_;

   $coro ||= $Coro::current;

   (Coro::async {
      Coro::State::trace $coro, 0;
      delete $coro->{_trace_sub_cb};
      delete $coro->{_trace_line_cb};
   })->prio (Coro::PRIO_MAX);

   Coro::cede;
}

sub ps_listing {
   my $times = Coro::State::enable_times;
   my $flags = $1;
   my $verbose = $flags =~ /v/;
   my $desc_format = $flags =~ /w/ ? "%-24s" : "%-24.24s";
   my $tim0_format = $times ? " %9s %8s " : " ";
   my $tim1_format = $times ? " %9.3f %8.3f " : " ";
   my $buf = sprintf "%20s %s%s %4s %4s$tim0_format$desc_format %s\n",
                     "PID", "S", "C", "RSS", "USES",
                     $times ? ("t_real", "t_cpu") : (),
                     "Description", "Where";
   for my $coro (reverse Coro::State::list) {
      my @bt;
      Coro::State::call ($coro, sub {
         # we try to find *the* definite frame that gives most useful info
         # by skipping Coro frames and pseudo-frames.
         for my $frame (1..10) {
            my @frame = caller $frame;
            @bt = @frame if $frame[2];
            last unless $bt[0] =~ /^Coro/;
         }
      });
      $bt[1] =~ s/^.*[\/\\]// if @bt && !$verbose;
      $buf .= sprintf "%20s %s%s %4s %4s$tim1_format$desc_format %s\n",
                      $coro+0,
                      $coro->is_new ? "N" : $coro->is_running ? "U" : $coro->is_ready ? "R" : "-",
                      $coro->is_traced ? "T" : $coro->has_cctx ? "C" : "-",
                      format_num4 $coro->rss,
                      format_num4 $coro->usecount,
                      $times ? $coro->times : (),
                      $coro->debug_desc,
                      (@bt ? sprintf "[%s:%d]", $bt[1], $bt[2] : "-");
   }

   $buf
}

=item command $string

Execute a debugger command, sending any output to STDOUT. Used by
C<session>, below.

=cut

sub command($) {
   my ($cmd) = @_;

   $cmd =~ s/\s+$//;

   if ($cmd =~ /^ps (?:\s* (\S+))? $/x) {
      print ps_listing;

   } elsif ($cmd =~ /^bt\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         my $bt;
         Coro::State::call ($coro, sub {
            local $Carp::CarpLevel = 2;
            $bt = eval { Carp::longmess "coroutine is" } || "$@";
         });
         if ($bt) {
            print $bt;
         } else {
            print "$1: unable to get backtrace\n";
         }
      }

   } elsif ($cmd =~ /^(?:e|eval)\s+(\d+)\s+(.*)$/) {
      if (my $coro = find_coro $1) {
         my $cmd = eval "sub { $2 }";
         my @res;
         Coro::State::call ($coro, sub { @res = eval { &$cmd } });
         print $@ ? $@ : (join " ", @res, "\n");
      }

   } elsif ($cmd =~ /^(?:tr|trace)\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         trace $coro;
      }

   } elsif ($cmd =~ /^(?:ut|untrace)\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         untrace $coro;
      }

   } elsif ($cmd =~ /^cancel\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         $coro->cancel;
      }

   } elsif ($cmd =~ /^ready\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         $coro->ready;
      }

   } elsif ($cmd =~ /^kill\s+(\d+)(?:\s+(.*))?$/) {
      my $reason = defined $2 ? $2 : "killed";

      if (my $coro = find_coro $1) {
         $coro->throw ($reason);
      }

   } elsif ($cmd =~ /^enable_times(\s+\S.*)?\s*$/) {
      my $enable = defined $1 ? 1*eval $1 : !Coro::State::enable_times;

      Coro::State::enable_times $enable;

      print "per-thread real and process time gathering ", $enable ? "enabled" : "disabled", ".\n";

   } elsif ($cmd =~ /^help$/) {
      print <<EOF;
ps [w|v]                show the list of all coroutines (wide, verbose)
bt <pid>                show a full backtrace of coroutine <pid>
eval <pid> <perl>       evaluate <perl> expression in context of <pid>
trace <pid>             enable tracing for this coroutine
untrace <pid>           disable tracing for this coroutine
kill <pid> <reason>	throws the given <reason> string in <pid>
cancel <pid>		cancels this coroutine
ready <pid>		force <pid> into the ready queue
enable_times <enable>	enable or disable time profiling in ps
<anything else>         evaluate as perl and print results
<anything else> &       same as above, but evaluate asynchronously
                        you can use (find_coro <pid>) in perl expressions
                        to find the coro with the given pid, e.g.
                        (find_coro 9768720)->ready
EOF

   } elsif ($cmd =~ /^(.*)&$/) {
      my $cmd = $1;
      my $sub = eval "sub { $cmd }";
      my $fh = select;
      Coro::async_pool {
         $Coro::current->{desc} = $cmd;
         my $t = Coro::Util::time;
         my @res = eval { &$sub };
         $t = Coro::Util::time - $t;
         print {$fh}
            "\rcommand: $cmd\n",
            "execution time: $t\n",
            "result: ", $@ ? $@ : (join " ", @res) . "\n",
            "> ";
      };

   } else {
      my @res = eval $cmd;
      print $@ ? $@ : (join " ", @res) . "\n";
   }

   local $| = 1;
}

=item session $fh

Run an interactive debugger session on the given filehandle. Each line entered
is simply passed to C<command> (with a few exceptions).

=cut

sub session($) {
   my ($fh) = @_;

   $fh = Coro::Handle::unblock $fh;
   my $old_fh = select $fh;
   my $guard = guard { select $old_fh };

   my $loglevel = $SESLOGLEVEL;
   local $log{$Coro::current} = sub {
      return unless $_[0] <= $loglevel;
      print $fh "\015", (format_msg $_[0], $_[1]), "> ";
   };

   print "coro debug session. use help for more info\n\n";

   while ((print "> "), defined (my $cmd = $fh->readline ("\012"))) {
      if ($cmd =~ /^exit\s*$/) {
         print "bye.\n";
         last;

      } elsif ($cmd =~ /^(?:ll|loglevel)\s*(\d+)?\s*/) {
         $loglevel = defined $1 ? $1 : -1;

      } elsif ($cmd =~ /^(?:w|watch)\s*([0-9.]*)\s+(.*)/) {
         my ($time, $cmd) = ($1*1 || 1, $2);
         my $cancel;

         Coro::async {
            $Coro::current->{desc} = "watch $cmd";
            select $fh;
            until ($cancel) {
               command $cmd;
               Coro::Timer::sleep $time;
            }
         };

         $fh->readable;
         $cancel = 1;

      } elsif ($cmd =~ /^help\s*/) {
         command $cmd;
         print <<EOF;
loglevel <int>		enable logging for messages of level <int> and lower
watch <time> <command>  repeat the given command until STDIN becomes readable
exit			end this session
EOF
      } else {
         command $cmd;
      }

      Coro::cede;
   }
}

=item $server = new_unix_server Coro::Debug $path

Creates a new unix domain socket that listens for connection requests and
runs C<session> on any connection. Normal unix permission checks and umask
applies, so you can protect your socket by puttint it into a protected
directory.

The C<socat> utility is an excellent way to connect to this socket:

   socat readline /path/to/socket

Socat also offers history support:

   socat readline:history=/tmp/hist.corodebug /path/to/socket

The server accepts connections until it is destroyed, so you must keep
the return value around as long as you want the server to stay available.

=cut

sub new_unix_server {
   my ($class, $path) = @_;

   unlink $path;
   my $unlink_guard = guard { unlink $path };

   AnyEvent::Socket::tcp_server "unix/", $path, sub {
      my ($fh) = @_;
      $unlink_guard; # mention it
      Coro::async_pool {
         $Coro::current->desc ("[Coro::Debug session]");
         session $fh;
      };
   } or Carp::croak "Coro::Debug::new_unix_server($path): $!";
}

=item $server = new_tcp_server Coro::Debug $port

Similar to C<new_unix_server>, but binds on a TCP port. I<Note that this is
usually results in a gaping security hole>.

Currently, only a TCPv4 socket is created, in the future, a TCPv6 socket
might also be created.

=cut

sub new_tcp_server {
   my ($class, $port) = @_;

   AnyEvent::Socket::tcp_server undef, $port, sub {
      my ($fh) = @_;
      Coro::async_pool {
         $Coro::current->desc ("[Coro::Debug session]");
         session $fh;
      };
   } or Carp::croak "Coro::Debug::new_tcp_server($port): $!";
}

sub DESTROY {
   my ($self) = @_;

   unlink $self->{path} if exists $self->{path};
   %$self = ();
}

1;

=back

=head1 AUTHOR/SUPPORT/CONTACT

   Marc A. Lehmann <schmorp@schmorp.de>
   http://software.schmorp.de/pkg/Coro.html

=cut


